go-controller/pkg/ovn/obj_retry.go

```diff
@@ -126,18 +189,1067 @@ func (r *retryObjs) getObjRetryEntry(key string) *retryObjEntry {
+			logicalSwitch: "host-networked",
+			name:          key,
+			uuid:          "host-networked",
+			ips:           []*net.IPNet{},
+			mac:           mac,
+		}
+	} else {
+		portInfo, _ = oc.logicalPortCache.get(key)
+	}
+	return portInfo
+}
+
+// Given an object and its type, getInternalCacheEntry returns the internal cache entry for this object.
+// This is now used only for pods, which will get their the logical port cache entry.
+func (oc *Controller) getInternalCacheEntry(objType reflect.Type, obj interface{}) interface{} {
+	switch objType {
+	case factory.PodType:
+		pod := obj.(*kapi.Pod)
+		return oc.getPortInfo(pod)
+	default:
+		return nil
+	}
+}
+
+// Given an object key and its type, getResourceFromInformerCache returns the latest state of the object
+// from the informers cache.
+func (oc *Controller) getResourceFromInformerCache(objType reflect.Type, key string) (interface{}, error) {
+	var obj interface{}
+	var err error
+
+	switch objType {
+	case factory.PolicyType:
+		namespace, name := splitNamespacedName(key)
+		obj, err = oc.watchFactory.GetNetworkPolicy(namespace, name)
+
+	case factory.NodeType:
+		obj, err = oc.watchFactory.GetNode(key)
+
+	case factory.PeerServiceTypeVar:
+		namespace, name := splitNamespacedName(key)
+		obj, err = oc.watchFactory.GetService(namespace, name)
+
+	case factory.PodType,
+		factory.PeerPodSelectorTypeVar,
+		factory.PeerPodForNamespaceAndPodSelectorType,
+		factory.LocalPodSelectorTypeVar:
+		namespace, name := splitNamespacedName(key)
+		obj, err = oc.watchFactory.GetPod(namespace, name)
+
+	case factory.PeerNamespaceAndPodSelectorTypeVar,
+		factory.PeerNamespaceSelectorTypeVar:
+		obj, err = oc.watchFactory.GetNamespace(key)
+
+	default:
+		err = fmt.Errorf("object type %v not supported, cannot retrieve it from informers cache",
+			objType)
+	}
+	return obj, err
+}
+
+// Given an object and its type, recordAddEvent records the add event on this object. Only used for pods now.
+func (oc *Controller) recordAddEvent(objType reflect.Type, obj interface{}) {
+	switch objType {
+	case factory.PodType:
+		klog.V(5).Infof("Recording add event on pod")
+		pod := obj.(*kapi.Pod)
+		oc.metricsRecorder.AddPod(pod.UID)
+	}
+}
+
+// Given an object and its type, recordDeleteEvent records the delete event on this object. Only used for pods now.
+func (oc *Controller) recordDeleteEvent(objType reflect.Type, obj interface{}) {
+	switch objType {
+	case factory.PodType:
+		klog.V(5).Infof("Recording add event on pod")
+		pod := obj.(*kapi.Pod)
+		oc.metricsRecorder.CleanPod(pod.UID)
+	}
+}
+
+// Given an object and its type, recordDeleteEvent records the delete event on this object. Only used for pods now.
+func (oc *Controller) recordErrorEvent(objType reflect.Type, obj interface{}, err error) {
+	switch objType {
+	case factory.PodType:
+		klog.V(5).Infof("Recording error event on pod")
+		pod := obj.(*kapi.Pod)
+		oc.metricsRecorder.AddPod(pod.UID)
+		oc.recordPodEvent(err, pod)
+	}
+}
```

Comment:
if this is unique to pods then keep it pod code doesn't need to be here right ?

Reply:
The idea is to make it easy in the future to record events for other resource types. This can be easily applied to egressfirewalls, for instance. I briefly showed this to @martinkennelly and he seemed to agree on the approach. He's also working on https://github.com/ovn-org/ovn-kubernetes/pull/2878, which adds more fields the to the retry entry struct and keeps track of the time it takes a pod to actually be correctly set up.